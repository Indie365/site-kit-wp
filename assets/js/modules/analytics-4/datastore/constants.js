/**
 * `modules/analytics-4` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const MODULES_ANALYTICS_4 = 'modules/analytics-4';

export const PROPERTY_CREATE = 'property_create';
export const WEBDATASTREAM_CREATE = 'webdatastream_create';

export const MAX_WEBDATASTREAMS_PER_BATCH = 10;

// Date range offset days for Analytics 4 report requests.
export const DATE_RANGE_OFFSET = 1;

export const GTM_SCOPE = 'https://www.googleapis.com/auth/tagmanager.readonly';

export const ENHANCED_MEASUREMENT_FORM = 'enhanced-measurement-form';
export const ENHANCED_MEASUREMENT_ENABLED = 'enhanced-measurement-enabled';
export const ENHANCED_MEASUREMENT_SHOULD_DISMISS_ACTIVATION_BANNER =
	'enhanced-measurement-should-dismiss-activation-banner';

export const FORM_CUSTOM_DIMENSIONS_CREATE = 'analytics4CustomDimensionsCreate';

export const CUSTOM_DIMENSION_DEFINITIONS = {
	googlesitekit_post_date: {
		parameterName: 'googlesitekit_post_date',
		displayName: __( 'WordPress Post Creation Date', 'google-site-kit' ),
		description: __(
			'Date of which this post was published',
			'google-site-kit'
		),
		scope: 'EVENT',
	},
	googlesitekit_post_author: {
		parameterName: 'googlesitekit_post_author',
		displayName: __( 'WordPress Post Author', 'google-site-kit' ),
		description: __(
			'User ID of the author for this post',
			'google-site-kit'
		),
		scope: 'EVENT',
	},
	googlesitekit_post_categories: {
		parameterName: 'googlesitekit_post_categories',
		displayName: __( 'WordPress Post Categories', 'google-site-kit' ),
		description: __(
			'Comma-separated list of category IDs assigned to this post',
			'google-site-kit'
		),
		scope: 'EVENT',
	},
	googlesitekit_post_type: {
		parameterName: 'googlesitekit_post_type',
		displayName: __( 'WordPress Post Type', 'google-site-kit' ),
		description: __( 'Content type for this post', 'google-site-kit' ),
		scope: 'EVENT',
	},
};
