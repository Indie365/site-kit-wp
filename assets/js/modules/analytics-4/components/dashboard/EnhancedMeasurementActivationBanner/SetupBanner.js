/**
 * EnhancedMeasurementActivationBanner > SetupBanner component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useState, useCallback, useEffect } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { SpinnerButton } from 'googlesitekit-components';
import { CORE_FORMS } from '../../../../../googlesitekit/datastore/forms/constants';
import { CORE_SITE } from '../../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../../googlesitekit/datastore/user/constants';
import {
	EDIT_SCOPE,
	FORM_SETUP,
} from '../../../../analytics/datastore/constants';
import {
	ENHANCED_MEASUREMENT_ENABLED,
	ENHANCED_MEASUREMENT_FORM,
	MODULES_ANALYTICS_4,
} from '../../../datastore/constants';
import { ERROR_CODE_MISSING_REQUIRED_SCOPE } from '../../../../../util/errors';
import {
	DAY_IN_SECONDS,
	getTimeInSeconds,
	trackEvent,
} from '../../../../../util';
import BannerNotification from '../../../../../components/notifications/BannerNotification';
import SuccessGreenSVG from '../../../../../../svg/graphics/ga4-success-green.svg';
import ErrorNotice from '../../../../../components/ErrorNotice';
import SurveyViewTrigger from '../../../../../components/surveys/SurveyViewTrigger';
import useViewContext from '../../../../../hooks/useViewContext';

const { useDispatch, useSelect } = Data;

export default function SetupBanner( props ) {
	const { isDismissed, onSubmitSuccess, onDismiss, children } = props;

	const viewContext = useViewContext();

	const [ errorNotice, setErrorNotice ] = useState( null );
	const [ isSaving, setIsSaving ] = useState( false );

	const hasEditScope = useSelect( ( select ) =>
		select( CORE_USER ).hasScope( EDIT_SCOPE )
	);

	const autoSubmit = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_SETUP, 'autoSubmit' )
	);

	const documentationURL = useSelect( ( select ) =>
		select( CORE_SITE ).getGoogleSupportURL( {
			path: '/analytics/answer/9216061',
		} )
	);

	const { submitChanges } = useDispatch( MODULES_ANALYTICS_4 );
	const { setPermissionScopeError } = useDispatch( CORE_USER );
	const { setValues } = useDispatch( CORE_FORMS );

	const commonSubmitChanges = useCallback( async () => {
		setIsSaving( true );

		setValues( ENHANCED_MEASUREMENT_FORM, {
			[ ENHANCED_MEASUREMENT_ENABLED ]: true,
		} );

		const { error } = await submitChanges();

		setIsSaving( false );

		if ( error ) {
			setErrorNotice( error );
			return;
		}

		// Ask the parent component to show the success banner.
		// This should be called last because it will unmount this component.
		onSubmitSuccess();
	}, [ setValues, submitChanges, onSubmitSuccess ] );

	const handleSubmitChanges = useCallback( async () => {
		const scopes = [];

		if ( hasEditScope === false ) {
			scopes.push( EDIT_SCOPE );
		}

		// If scope not granted, trigger scope error right away. These are
		// typically handled automatically based on API responses, but
		// this particular case has some special handling to improve UX.
		if ( scopes.length > 0 ) {
			setValues( FORM_SETUP, { autoSubmit: true } );

			setPermissionScopeError( {
				code: ERROR_CODE_MISSING_REQUIRED_SCOPE,
				message: __(
					'Additional permissions are required to enable enhanced measurement for the selected web data stream.',
					'google-site-kit'
				),
				data: {
					status: 403,
					scopes,
					skipModal: true,
					redirectURL: global.location.href,
				},
			} );
			return;
		}

		trackEvent(
			`${ viewContext }_enhanced-measurement-notification`,
			'confirm_notification'
		);

		await commonSubmitChanges();
	}, [
		viewContext,
		hasEditScope,
		commonSubmitChanges,
		setPermissionScopeError,
		setValues,
	] );

	const handleDismiss = useCallback( () => {
		trackEvent(
			`${ viewContext }_enhanced-measurement-notification`,
			'dismiss_notification'
		);

		onDismiss?.();
	}, [ onDismiss, viewContext ] );

	const handleLearnMore = useCallback( () => {
		trackEvent(
			`${ viewContext }_enhanced-measurement-notification`,
			'click_learn_more_link'
		);
	}, [ viewContext ] );

	useEffect( () => {
		trackEvent(
			`${ viewContext }_enhanced-measurement-notification`,
			'view_notification'
		);
	}, [ viewContext ] );

	// If the user lands back on this component with autoSubmit and the edit scope,
	// resubmit the form.
	useEffect( () => {
		async function handleAutoSubmit() {
			// Auto-submit should only auto-invoke once.
			setValues( FORM_SETUP, { autoSubmit: false } );

			await commonSubmitChanges();
		}

		if ( autoSubmit && hasEditScope ) {
			handleAutoSubmit();
		}
	}, [
		hasEditScope,
		setValues,
		commonSubmitChanges,
		onSubmitSuccess,
		autoSubmit,
	] );

	const description = hasEditScope
		? __(
				'Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required.',
				'google-site-kit'
		  )
		: __(
				'Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required — you’ll be redirected to give permission for Site Kit to enable it on your behalf.',
				'google-site-kit'
		  );

	return (
		<BannerNotification
			id="googlesitekit-enhanced-measurement-activation-banner"
			className="googlesitekit-enhanced-measurement-setup-banner"
			title={ __(
				'Understand how visitors interact with your content',
				'google-site-kit'
			) }
			description={ description }
			learnMoreLabel={ __( 'Learn more', 'google-site-kit' ) }
			learnMoreURL={ documentationURL }
			onLearnMoreClick={ handleLearnMore }
			ctaComponent={
				<SpinnerButton
					onClick={ handleSubmitChanges }
					isSaving={ isSaving }
				>
					{ __( 'Enable now', 'google-site-kit' ) }
				</SpinnerButton>
			}
			footer={
				<p className="googlesitekit-enhanced-measurement-setup-banner__footer-notice">
					{ __(
						'You can always add/edit this in the Site Kit Settings',
						'google-site-kit'
					) }
				</p>
			}
			dismiss={ __( 'Maybe later', 'google-site-kit' ) }
			WinImageSVG={ () => <SuccessGreenSVG /> }
			format="small"
			// Although the banner does handle its own dismiss state via a dismissable item, we still need to
			// provide a value here to ensure BannerNotification's own dismiss state is expired.
			dismissExpires={ getTimeInSeconds( 'month' ) }
			onDismiss={ handleDismiss }
		>
			{ errorNotice && <ErrorNotice error={ errorNotice } /> }
			{ ! isDismissed && (
				<SurveyViewTrigger
					triggerID="view_enhanced_measurement_cta"
					ttl={ DAY_IN_SECONDS }
				/>
			) }
			{ children }
		</BannerNotification>
	);
}

SetupBanner.propTypes = {
	onSubmitSuccess: PropTypes.func,
	onDismiss: PropTypes.func,
	children: PropTypes.node,
};
