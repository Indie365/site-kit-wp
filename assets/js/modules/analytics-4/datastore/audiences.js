/**
 * `modules/analytics-4` data store: audiences.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import Data from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from './constants';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { createValidatedAction } from '../../../googlesitekit/data/utils';
import { validateAudience } from '../utils/validation';

const fetchGetAudiencesStore = createFetchStore( {
	baseName: 'getAudiences',
	controlCallback() {
		return API.get(
			'modules',
			'analytics-4',
			'audiences',
			{},
			{
				useCache: false,
			}
		);
	},
	reducerCallback( state, audiencesResponse ) {
		return { ...state, audiences: audiencesResponse.audiences };
	},
} );

const fetchCreateAudienceStore = createFetchStore( {
	baseName: 'createAudience',
	controlCallback: ( { audience } ) =>
		API.set( 'modules', 'analytics-4', 'create-audience', {
			audience,
		} ),
	reducerCallback( state, audience ) {
		return {
			...state,
			audiences: [ ...( state.audiences || [] ), audience ],
		};
	},
	argsToParams: ( audience ) => ( {
		audience,
	} ),
	validateParams: ( { audience } ) => {
		validateAudience( audience );
	},
} );

const baseInitialState = {
	audiences: undefined,
};

const baseActions = {
	/**
	 * Creates new property audience.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} audience                             The property audience parameters.
	 * @param {string} [audience.displayName]               Required. The display name of the Audience.
	 * @param {string} [audience.description]               Required. The description of the Audience.
	 * @param {number} [audience.membershipDurationDays]    Required. The duration a user should stay in an Audience. Cannot be more than 540 days.
	 * @param {Array}  [audience.filterClauses]             Required. Filter clauses array of <AudienceFilterClause> objects that define the Audience.
	 * @param {Object} [audience.eventTrigger]              Optional. Specifies an event to log when a user joins the Audience.
	 * @param {string} [audience.eventTrigger.eventName]    Required if `eventTrigger` is provided. The event name that will be logged.
	 * @param {string} [audience.eventTrigger.logCondition] Required if `eventTrigger` is provided. When to log the event. Acceptable values:
	 *                                                      - 'LOG_CONDITION_UNSPECIFIED': Log condition is not specified.
	 *                                                      - 'AUDIENCE_JOINED': The event should be logged only when a user is joined.
	 *                                                      - 'AUDIENCE_MEMBERSHIP_RENEWED': The event should be logged whenever the Audience condition is met, even if the user is already a member of the Audience.
	 * @param {string} [audience.exclusionDurationMode]     Optional. Specifies how long an exclusion lasts for users that meet the exclusion filter. Acceptable values:
	 *                                                      - 'AUDIENCE_EXCLUSION_DURATION_MODE_UNSPECIFIED': Not specified.
	 *                                                      - 'EXCLUDE_TEMPORARILY': Exclude users from the Audience during periods when they meet the filter clause.
	 *                                                      - 'EXCLUDE_PERMANENTLY': Exclude users from the Audience if they've ever met the filter clause.
	 * @return {Object} Object with `response` and `error`.
	 */
	createAudience: createValidatedAction(
		validateAudience,
		function* ( audience ) {
			const { response, error } =
				yield fetchCreateAudienceStore.actions.fetchCreateAudience(
					audience
				);

			return { response, error };
		}
	),
};

const baseControls = {};

const baseReducer = ( state, { type } ) => {
	switch ( type ) {
		default: {
			return state;
		}
	}
};

const baseResolvers = {
	*getAudiences() {
		const registry = yield Data.commonActions.getRegistry();

		const audiences = registry.select( MODULES_ANALYTICS_4 ).getAudiences();

		if ( audiences === undefined ) {
			yield fetchGetAudiencesStore.actions.fetchGetAudiences();
		}
	},
};

const baseSelectors = {
	/**
	 * Gets the property audiences.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} state Data store's state.
	 * @return {(Array|undefined)} An array with audiences objects; `undefined` if not loaded.
	 */
	getAudiences( state ) {
		return state.audiences;
	},
};

const store = Data.combineStores(
	fetchGetAudiencesStore,
	fetchCreateAudienceStore,
	{
		initialState: baseInitialState,
		actions: baseActions,
		controls: baseControls,
		reducer: baseReducer,
		resolvers: baseResolvers,
		selectors: baseSelectors,
	}
);

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
