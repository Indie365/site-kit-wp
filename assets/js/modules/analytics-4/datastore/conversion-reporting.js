/**
 * `modules/analytics-4` data store: conversion reporting.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { createRegistrySelector } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from './constants';

export const selectors = {
	/**
	 * Checks whether the provided conversion reporting events are available.
	 *
	 * @since 1.135.0
	 *
	 * @param {Object}               state  Data store's state.
	 * @param {string|Array<string>} events Conversion reporting events to check.
	 * @return {(boolean|undefined)} True if all provided custom dimensions are available, otherwise false. Undefined if available custom dimensions are not loaded yet.
	 */
	hasConversionReportingEvents: createRegistrySelector(
		( select ) => ( state, events ) => {
			// Ensure events is always an array, even if a string is passed.
			const eventsToCheck = Array.isArray( events ) ? events : [ events ];

			const detectedEvents =
				select( MODULES_ANALYTICS_4 ).getDetectedEvents();

			if ( ! detectedEvents?.length ) {
				return false;
			}

			return eventsToCheck.some( ( event ) =>
				detectedEvents.includes( event )
			);
		}
	),
};

export default {
	selectors,
};
