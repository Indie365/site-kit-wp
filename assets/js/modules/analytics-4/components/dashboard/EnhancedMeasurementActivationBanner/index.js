/**
 * EnhancedMeasurementActivationBanner component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { useCallback, useEffect, useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { CORE_MODULES } from '../../../../../googlesitekit/modules/datastore/constants';
import { CORE_USER } from '../../../../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../../datastore/constants';
import {
	ACTIVATION_STEP_SETUP,
	ACTIVATION_STEP_SUCCESS,
	ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY,
	ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
} from '../../../constants';
import { useTooltipState } from '../../../../../components/AdminMenuTooltip/useTooltipState';
import { useShowTooltip } from '../../../../../components/AdminMenuTooltip/useShowTooltip';
import { AdminMenuTooltip } from '../../../../../components/AdminMenuTooltip/AdminMenuTooltip';
import SetupBanner from './SetupBanner';
import SuccessBanner from './SuccessBanner';
import { getTimeInSeconds } from '../../../../../util';
import whenActive from '../../../../../util/when-active';
import {
	isValidPropertyID,
	isValidWebDataStreamID,
} from '../../../utils/validation';

const { useSelect, useDispatch } = Data;

function EnhancedMeasurementActivationBanner() {
	const [ step, setStep ] = useState( ACTIVATION_STEP_SETUP );
	const [
		isEnhancedMeasurementInitiallyDisabled,
		setIsEnhancedMeasurementInitiallyDisabled,
	] = useState( undefined );

	const propertyID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getPropertyID()
	);

	const webDataStreamID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getWebDataStreamID()
	);

	const isBannerDismissed = useSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed(
			ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY
		)
	);

	const hasAnalytics4Access = useSelect( ( select ) =>
		select( CORE_MODULES ).hasModuleOwnershipOrAccess( 'analytics-4' )
	);

	const isEnhancedMeasurementStreamEnabled = useSelect( ( select ) => {
		if (
			! isValidPropertyID( propertyID ) ||
			! isValidWebDataStreamID( webDataStreamID ) ||
			! hasAnalytics4Access ||
			isBannerDismissed
		) {
			return undefined;
		}

		return select( MODULES_ANALYTICS_4 ).isEnhancedMeasurementStreamEnabled(
			propertyID,
			webDataStreamID
		);
	} );

	const { dismissItem } = useDispatch( CORE_USER );

	const { isTooltipVisible } = useTooltipState(
		ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY
	);

	const showTooltip = useShowTooltip(
		ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY
	);

	function handleDismiss() {
		showTooltip();
		dismissItem(
			ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
			{
				expiresInSeconds: getTimeInSeconds( 'month' ),
			}
		);
	}

	const handleSubmit = useCallback( () => {
		if ( step === ACTIVATION_STEP_SETUP ) {
			setStep( ACTIVATION_STEP_SUCCESS );
		}
	}, [ step ] );

	useEffect( () => {
		if (
			isEnhancedMeasurementStreamEnabled === false &&
			isEnhancedMeasurementInitiallyDisabled === undefined
		) {
			setIsEnhancedMeasurementInitiallyDisabled( true );
		}
	}, [
		isEnhancedMeasurementInitiallyDisabled,
		isEnhancedMeasurementStreamEnabled,
	] );

	if ( isTooltipVisible ) {
		return (
			<AdminMenuTooltip
				title={ __(
					'Enable enhanced measurement later here',
					'google-site-kit'
				) }
				content={ __(
					'You can always turn on enhanced measurement later in Site Kit Settings.',
					'google-site-kit'
				) }
				dismissLabel={ __( 'Got it', 'google-site-kit' ) }
				tooltipStateKey={
					ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY
				}
			/>
		);
	}

	if ( ! isEnhancedMeasurementInitiallyDisabled || isBannerDismissed ) {
		return null;
	}

	if ( step === ACTIVATION_STEP_SETUP ) {
		return (
			<SetupBanner
				onSubmitSuccess={ handleSubmit }
				onDismiss={ handleDismiss }
			/>
		);
	}

	if ( step === ACTIVATION_STEP_SUCCESS ) {
		return <SuccessBanner />;
	}

	return null;
}

export default whenActive( { moduleName: 'analytics-4' } )(
	EnhancedMeasurementActivationBanner
);
