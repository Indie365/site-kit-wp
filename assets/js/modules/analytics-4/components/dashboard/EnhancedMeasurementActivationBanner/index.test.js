/**
 * EnhancedMeasurementActivationBanner component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	render,
	provideModules,
	provideModuleRegistrations,
	provideUserInfo,
	provideUserAuthentication,
	fireEvent,
	waitFor,
} from '../../../../../../../tests/js/test-utils';
import { CORE_MODULES } from '../../../../../googlesitekit/modules/datastore/constants';
import { CORE_USER } from '../../../../../googlesitekit/datastore/user/constants';
import { EDIT_SCOPE } from '../../../../analytics/datastore/constants';
import { MODULES_ANALYTICS_4 } from '../../../datastore/constants';
import { ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY } from '../../../constants';
import EnhancedMeasurementActivationBanner from './index';

describe( 'EnhancedMeasurementActivationBanner', () => {
	const propertyID = '1000';
	const webDataStreamID = '2000';

	let enhancedMeasurementSettingsMock;
	let registry;

	beforeEach( () => {
		enhancedMeasurementSettingsMock = {
			fileDownloadsEnabled: null,
			name: 'properties/1000/dataStreams/2000/enhancedMeasurementSettings',
			outboundClicksEnabled: null,
			pageChangesEnabled: null,
			scrollsEnabled: null,
			searchQueryParameter: 'q,s,search,query,keyword',
			siteSearchEnabled: null,
			streamEnabled: false,
			uriQueryParameter: null,
			videoEngagementEnabled: null,
		};

		registry = createTestRegistry();

		provideUserInfo( registry );
		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideModuleRegistrations( registry );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			propertyID,
			webDataStreamID,
			ownerID: 1,
		} );

		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				enhancedMeasurementSettingsMock,
				{ propertyID, webDataStreamID }
			);
	} );

	it( 'should render the setup step when enhanced measurement is initially false and the banner is not dismissed', async () => {
		const { container, getByRole } = render(
			<EnhancedMeasurementActivationBanner />,
			{
				registry,
			}
		);

		await waitFor( () => expect( container ).toMatchSnapshot() );

		await waitFor( () =>
			expect(
				getByRole( 'button', { name: 'Enable now' } )
			).toBeInTheDocument()
		);
	} );

	it( 'should render the success step when the the setup form is successfully submitted', async () => {
		const { container, getByRole, getByText, waitForRegistry } = render(
			<EnhancedMeasurementActivationBanner />,
			{
				registry,
			}
		);

		fireEvent.click( getByRole( 'button', { name: 'Enable now' } ) );

		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect(
			getByText(
				'You successfully enabled enhanced measurement for your site'
			)
		).toBeInTheDocument();
	} );

	it.each( [
		[
			'there is not a valid propertyID',
			() => {
				registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
					propertyID: '',
				} );
			},
		],
		[
			'there is not a valid webDataStreamID',
			() => {
				registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
					webDataStreamID: '',
				} );
			},
		],
		[
			'the user does not have access to the Analytics 4 module',
			() => {
				registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
					ownerID: 2,
				} );

				registry
					.dispatch( CORE_MODULES )
					.receiveCheckModuleAccess(
						{ access: false },
						{ slug: 'analytics-4' }
					);
			},
		],
		[
			'enhanced measurement is initially true',
			() => {
				registry
					.dispatch( MODULES_ANALYTICS_4 )
					.receiveGetEnhancedMeasurementSettings(
						{
							...enhancedMeasurementSettingsMock,
							streamEnabled: true,
						},
						{ propertyID, webDataStreamID }
					);
			},
		],
		[
			'the banner is dismissed',
			() => {
				registry
					.dispatch( CORE_USER )
					.receiveGetDismissedItems( [
						ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
					] );
			},
		],
	] )( 'should not render when %s', ( _, setupTestCase ) => {
		setupTestCase();

		const { container } = render( <EnhancedMeasurementActivationBanner />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );
} );
