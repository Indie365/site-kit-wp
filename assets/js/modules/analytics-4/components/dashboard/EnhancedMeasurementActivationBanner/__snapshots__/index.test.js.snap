// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnhancedMeasurementActivationBanner should render the setup step when enhanced measurement is initially false and the banner is not dismissed 1`] = `
<div>
  <section
    class="googlesitekit-enhanced-measurement-setup-banner googlesitekit-publisher-win--small googlesitekit-publisher-win--is-open googlesitekit-publisher-win"
    id="googlesitekit-enhanced-measurement-activation-banner"
  >
    <div
      class="mdc-layout-grid"
    >
      <div
        class="mdc-layout-grid__inner"
      >
        <div
          class="googlesitekit-publisher-win__content mdc-layout-grid__cell mdc-layout-grid__cell--span-8-desktop mdc-layout-grid__cell--span-6-tablet mdc-layout-grid__cell--span-4-phone"
        >
          <div
            class="googlesitekit-publisher-win__title-image-wrapper"
          >
            <h3
              class="googlesitekit-heading-2 googlesitekit-publisher-win__title"
            >
              Understand how visitors interact with your content
            </h3>
            <div
              class="googlesitekit-publisher-win__image-small"
            >
              <svg />
            </div>
          </div>
          <div
            class="googlesitekit-publisher-win__desc"
          >
            <p>
              <span>
                Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required.
              </span>
               
              <a
                aria-label="Learn more (opens in a new tab)"
                class="googlesitekit-cta-link googlesitekit-cta-link--external"
                href="https://support.google.com/analytics/answer/9216061?hl=en-US"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more
              </a>
            </p>
          </div>
          <div
            class="googlesitekit-publisher-win__actions"
          >
            <button
              class="mdc-button googlesitekit-button-icon--spinner mdc-button--raised"
              target="_self"
            >
              <span
                class="mdc-button__label"
              >
                Enable now
              </span>
            </button>
            <button
              class="googlesitekit-cta-link"
            >
              Maybe later
            </button>
          </div>
          <div
            class="googlesitekit-publisher-win__footer"
          >
            <p
              class="googlesitekit-enhanced-measurement-setup-banner__footer-notice"
            >
              You can always add/edit this in the Site Kit Settings
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`EnhancedMeasurementActivationBanner should render the success step when the the setup form is successfully submitted 1`] = `
<div>
  <section
    class="googlesitekit-enhanced-measurement-success-banner googlesitekit-publisher-win--small googlesitekit-publisher-win--win-success googlesitekit-publisher-win--is-open googlesitekit-publisher-win"
    id="googlesitekit-enhanced-measurement-activation-banner"
  >
    <div
      class="mdc-layout-grid"
    >
      <div
        class="mdc-layout-grid__inner"
      >
        <div
          class="googlesitekit-publisher-win__content mdc-layout-grid__cell mdc-layout-grid__cell--span-8-desktop mdc-layout-grid__cell--span-6-tablet mdc-layout-grid__cell--span-4-phone"
        >
          <div
            class="googlesitekit-publisher-win__title-image-wrapper"
          >
            <h3
              class="googlesitekit-heading-2 googlesitekit-publisher-win__title"
            >
              You successfully enabled enhanced measurement for your site
            </h3>
            <div
              class="googlesitekit-publisher-win__image-small"
            >
              <svg />
            </div>
          </div>
          <div
            class="googlesitekit-publisher-win__desc"
          >
            <p>
              <span>
                Your configured Analytics web data stream will now automatically measure interactions on your site in addition to standard page views measurement.
              </span>
               
              <a
                aria-label="Learn more (opens in a new tab)"
                class="googlesitekit-cta-link googlesitekit-cta-link--external"
                href="https://support.google.com/analytics/answer/9216061?hl=en-US"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more
              </a>
            </p>
          </div>
          <div
            class="googlesitekit-publisher-win__actions"
          >
            <button
              class="mdc-button mdc-button--raised"
              target="_self"
            >
              <span
                class="mdc-button__label"
              >
                OK, Got it
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
